src/main/java/com/segment/analytics/android/integrations/firebase/FirebaseIntegration.java

```diff
@@ -3,88 +3,242 @@
-
-    if (!isNullOrEmpty(identify.userId())) {
-      mFirebaseAnalytics.setUserId(identify.userId());
+  private String mapEvent(String event) {
+    String eventName = event;
+    if (eventMapper.containsKey(eventName)) {
+      eventName = eventMapper.get(eventName);
     }
+    eventName = eventName.trim().replaceAll(" ", "_");
+    if (eventName.length() > 40) {
+      eventName = trimKey(eventName);
+    }
+    return eventName;
+  }
 
-    // mFirebaseAnalytics.setUserProperty(key, value);
-
+  private Bundle formatProperties(Properties properties) {
+    Bundle bundle = new Bundle();
+    if ((properties.revenue() != 0 || properties.total() != 0)
+        && isNullOrEmpty(properties.currency())) {
+      bundle.putString(Param.CURRENCY, "USD");
+    }
+    for (Map.Entry<String, Object> entry : properties.entrySet()) {
+      String property = entry.getKey();
+      property = mapProperty(property);
+      if (entry.getValue() instanceof Integer) {
+        int value = (int) entry.getValue();
+        bundle.putInt(property, value);
+        logger.verbose("bundle.putInt(%s, %s);", property, value);
+        continue;
+      }
+      if (entry.getValue() instanceof Double) {
+        double value = (double) entry.getValue();
+        bundle.putDouble(property, value);
+        logger.verbose("bundle.putDouble(%s, %s);", property, value);
+        continue;
+      }
+      if (entry.getValue() instanceof Long) {
+        long value = (long) entry.getValue();
+        bundle.putLong(property, value);
+        logger.verbose("bundle.putLong(%s, %s);", property, value);
+        continue;
+      }
+      if (entry.getValue() instanceof String) {
+        String value = String.valueOf(entry.getValue());
+        bundle.putString(property, value);
+        logger.verbose("bundle.putString(%s, %s);", property, value);
+        continue;
+      }
+      if (entry.getValue() instanceof Date) {
+        Date value = (Date) entry.getValue();
+        String formattedDate = formatDate(value);
+        bundle.putString(property, formattedDate);
+        logger.verbose("bundle.putString(%s, %s);", property, formattedDate);
+        continue;
+      }
+    }
+    return bundle;
   }
 
-  @Override public void screen(ScreenPayload screen) {
-    super.screen(screen);
-//    mFirebaseAnalytics.setCurrentScreen(this, screen.name(), null /* class override */);
+  private String mapProperty(String property) {
+    if (propertyMapper.containsKey(property)) {
+      property = propertyMapper.get(property);
+    }
+    property = property.trim().replaceAll(" ", "_");
+    if (property.length() > 40) {
+      property = trimKey(property);
+    }
+    return property;
   }
 
-  @Override public void track(TrackPayload track) {
-    super.track(track);
+  private String trimKey(String string) {
+    return string.substring(0, Math.min(string.length(), 40));
+  }
 
+  private String formatDate(Date date) {
+    String stringifiedValue = toISO8601Date(date);
```

Comment:
We only need the first 10 characters so this is wasteful. This should use a date formatter directly to produce the stringified version.

Reply:
Were you thinking something like SimpleDateFormatter? 

```
import java.text.SimpleDateFormat;

SimpleDateFormat firebaseFormat = new SimpleDateFormat("yyyy-MM-dd");
String formattedDate = firebaseFormat.format(date);
```

Reply:
Yep. The `firebaseFormat` variable should be created once as a `private static final SimpleDateFormat FIREBASE_FORMAT = new SimpleDateFormat("yyyy-MM-dd");` and re-used rather than created inside this function.